### PR TITLE
(GH-20) Add cChocoConfig resource

### LINKED ISSUE #20: Add cChocoConfig Resource
I have had some issues installing 32 bit Choco packages on 64 bit machines with SYSTEM user due to incorrect temp path resolution. 
There should be a way to configure cacheLocation in cChocoInstaller using DSC. Choco introduced this in v0.9.9.9 with "choco config set cacheLocation fullPath"
 https://github.com/chocolatey/choco/wiki/How-To-Change-Cache

### PR DESCRIPTION
**Pull request by mkevenaar**
This resource allows users to change Chocolatey Config
settings using an DSC resource.

Fixes #20 

replaces PR #132

### CODE REVIEW

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        [parameter(Mandatory = $true)]
+        [System.String]
+        $ConfigName,
+
+        [ValidateSet('Present','Absent')]
+        [string]
+        $Ensure='Present',
+
+        [parameter(Mandatory = $false)]
+        [System.String]
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we just use `[string]` here?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+Set-TargetResource -ConfigName cacheLocation -Ensure 'Present' -Value 'c:\temp\choco'
+
+#>
+function Set-TargetResource
+{
+    [CmdletBinding(SupportsShouldProcess=$true)]
+    param
+    (
+        [parameter(Mandatory = $true)]
+        [System.String]
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we just use `[string]` here?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+Get-TargetResource -ConfigName cacheLocation -Ensure 'Present' -Value 'c:\temp\choco'
+#>
+function Get-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Collections.Hashtable])]
+    param
+    (
+        [parameter(Mandatory = $true)]
+        [System.String]
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we just use `[string]` here?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        $ConfigName,
+
+        [ValidateSet('Present','Absent')]
+        [System.String]
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we just use `[string]` here?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        $Ensure='Present',
+
+        [parameter(Mandatory = $false)]
+        [System.String]
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we just use `[string]` here?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+
+    Write-Verbose "Starting cChocoConfig Set-TargetResource - Config Name: $ConfigName, Ensure: $Ensure"
+
+    if ($pscmdlet.ShouldProcess("Choco config $ConfigName will be ensured $Ensure."))
+    {
+        if ($Ensure -eq 'Present')
+        {
+            Write-Verbose "Setting choco config $ConfigName."
+            choco config set --name $ConfigName --value $Value
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: What if `$value` is empty? We don't appear to be testing for that.

Can we also use the [Scripting Best Practices](https://chocolatey.org/docs/commands-reference#scripting-integration---best-practices-style-guide) and use `--name="'$ConfigName'" --value="'$Value'"` ?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        }
+        else 
+        {
+            Write-Verbose "Unsetting choco config $ConfigName."
+            choco config unset --name $ConfigName
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Amend to use Scripting Best Practices?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+.Description
+Performs the test for cChocoFeature.
+
+.Example
+Test-TargetResource -ConfigName cacheLocation -Ensure 'Present' -Value 'c:\temp\choco'
+#>
+function Test-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Boolean])]
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we just use `Boolean` or `Bool` here?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+.Description
+Returns the configuration for cChocoConfig.
+
+.Example
+Get-TargetResource -ConfigName cacheLocation -Ensure 'Present' -Value 'c:\temp\choco'
+#>
+function Get-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Collections.Hashtable])]
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we just use `Hashtable` here?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    Write-Verbose "Starting cChocoConfig Test-TargetResource - Config Name: $ConfigName, Ensure: $Ensure."
+
+    # validate value is given when ensure present
+    if ($Ensure -eq 'Present' -and -not $PSBoundParameters.ContainsKey('Value')) {
+        throw "Missing parameter 'Value' when ensuring config is present!"
+    }
+
+	if($env:ChocolateyInstall -eq "" -or $null -eq $env:ChocolateyInstall)
+	{
+		$exe = (get-command choco).Source
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we capitalise the `Get-Command` and use the full parameter names?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+		$chocofolder = $exe.Substring(0,$exe.LastIndexOf("\"))
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Is there a reason we can't use the Split-Path cmdlets here (for simplication which I always prefer)? 

Can you also do some testing to make sure you are getting the results you need? What happens for example if `Get-Command -Command choco` returns nothing or throws? Same with `$chocofolder`?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+		{
+			$chocofolder = $chocofolder.Substring(0,$chocofolder.LastIndexOf("\"))
+		}
+	}
+	else
+	{
+		$chocofolder = $env:ChocolateyInstall
+	}
+	$configfolder = "$chocofolder\config"
+	$configfile = Get-ChildItem $configfolder | Where-Object {$_.Name -match "chocolatey.config$"}
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can we use the full parameter names for `Get-ChildItem`?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+    # There is currently no choco command that only returns the settings in an CSV format.
+    # choco config list -r shows settings, sources, features and a note about API keys.
+    $xml = [xml](Get-Content $configfile.FullName)
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Full parameter names for `Get-Content`? What do we do if the file doesn't exist or cannot be read? We're not checking for that.

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+	if($Ensure -eq 'Present')
+	{
+		return $false
+	}
+	else
+	{
+		return $true
+	}
+}
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Formatting is a best messed up here.

File: DSCResources/cChocoConfig/cChocoConfig.psm1 [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+		{
+            return ($setting.value -eq $Value)
+		}
+		elseif($setting.key -eq $ConfigName -and $Ensure -eq 'Absent')
+		{
+            return ([String]::IsNullOrEmpty($setting.value))
+		}
+	}
+
+	if($Ensure -eq 'Present')
```
Comment (pauby) [inline DSCResources/cChocoConfig/cChocoConfig.psm1]: Can you add some comments here to clarify this (I understand it, I just think it needs some explanation for future)?

**Review by pauby — CHANGES_REQUESTED** [review]
Just some minor changes to be made.

Can I ask for a few more comments in the code as well?

File: DSCResources/cChocoConfig/cChocoConfig.psm1 (hunk 6 above)
Comment (mkevenaar) [reply DSCResources/cChocoConfig/cChocoConfig.psm1]: I am checking the value of `$value` here: https://github.com/chocolatey/cChoco/pull/145/files#diff-6ff49aeca70224dbfb6b7427636392dfR127

**Comment by mkevenaar** [comment]
@pauby I have updated the code with your feedback.

File: DSCResources/cChocoConfig/cChocoConfig.psm1 (hunk 6 above)
Comment (pauby) [reply DSCResources/cChocoConfig/cChocoConfig.psm1]: I may be missing it but all that is being checked there is that the parameter `Value` has been passed. There is no check that it's empty or `$null`?

**Review by pauby — COMMENTED** [review]
This awesome. Love the comments. Just one more question left.

File: DSCResources/cChocoConfig/cChocoConfig.psm1 (hunk 6 above)
Comment (mkevenaar) [reply DSCResources/cChocoConfig/cChocoConfig.psm1]: I have updated this just now and added tests for an empty string and $null

File: AppVeyor/AppVeyorInstall.ps1 [hunk 16]
```diff
@@ -29,8 +34,9 @@ Write-Host "Installed NuGet version '$($pkg.version)'"
 # Install Modules                 #
 #---------------------------------#
 [version]$ScriptAnalyzerVersion = '1.8.1'
+[version]$PesterVersion = '4.10.1'
 Install-Module -Name 'PSScriptAnalyzer' -Repository PSGallery -Force -ErrorAction Stop -MaximumVersion $ScriptAnalyzerVersion
-Install-Module -Name 'Pester' -SkipPublisherCheck -Repository PSGallery -Force -ErrorAction Stop
+Install-Module -Name 'Pester' -SkipPublisherCheck -Repository PSGallery -Force -ErrorAction Stop -MaximumVersion $PesterVersion
```
Comment (pauby) [inline AppVeyor/AppVeyorInstall.ps1:39]: Why are we setting this here? What do later versions break?

Comment (mkevenaar) [reply AppVeyor/AppVeyorInstall.ps1:39]: The version 5.x of Pester has a lot of breaking changes. I was unable to get Pester 5 working with the current code. 

For breaking changes see: https://github.com/pester/Pester/releases/tag/5.0.0#breaking-changes

**Comment by pauby** [comment]
Awesome work! Thanks for your contribution!

**Pull request merged**